mm2src/mm2_net/src/wasm/body_stream.rs

```diff
@@ -0,0 +1,411 @@
+    CompressionFlag,
+    DataLength,
+    Data(u32),
+    TrailerLength,
+    Trailer(u32),
+    Done,
+}
+
+impl ReadState {
+    fn is_done(&self) -> bool { matches!(self, ReadState::Done) }
+
+    fn finished_data(&self) -> bool {
+        matches!(self, ReadState::TrailerLength)
+            || matches!(self, ReadState::Trailer(_))
+            || matches!(self, ReadState::Done)
+    }
+}
+
+/// Handles the HTTP response body, decoding data, and extracting trailers
+#[pin_project]
+pub struct ResponseBody {
+    #[pin]
+    body_stream: BodyStream,
+    buf: EncodedBytes,
+    incomplete_data: BytesMut,
+    data: Option<BytesMut>,
+    trailer: Option<HeaderMap>,
+    state: ReadState,
+    finished_stream: bool,
+}
+
+impl ResponseBody {
+    /// Creates a new `ResponseBody` based on a ReadableStream and content type.
+    pub(crate) async fn new(body_stream: ReadableStream, content_type: &str) -> Result<Self, PostGrpcWebErr> {
+        let body_stream: ReadableStreamDefaultReader = body_stream
+            .get_reader()
+            .dyn_into()
+            .map_err(|err| PostGrpcWebErr::BadResponse(format!("{err:?}")))?;
+
+        Ok(Self {
+            body_stream: BodyStream::new(body_stream).await?,
+            buf: EncodedBytes::new(content_type)?,
+            incomplete_data: BytesMut::new(),
+            data: None,
+            trailer: None,
+            state: ReadState::CompressionFlag,
+            finished_stream: false,
+        })
+    }
+
+    fn read_stream(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), PostGrpcWebErr>> {
+        if self.finished_stream {
+            return Poll::Ready(Ok(()));
+        }
+
+        let this = self.project();
+
+        match ready!(this.body_stream.poll_data(cx)) {
+            Some(Ok(data)) => {
+                if let Err(e) = this.buf.append(data) {
+                    return Poll::Ready(Err(e));
+                }
+
+                Poll::Ready(Ok(()))
+            },
+            Some(Err(e)) => Poll::Ready(Err(e)),
+            None => {
+                *this.finished_stream = true;
+                Poll::Ready(Ok(()))
+            },
+        }
+    }
+
+    fn step(self: Pin<&mut Self>) -> Result<(), PostGrpcWebErr> {
+        let this = self.project();
+
+        loop {
+            match this.state {
+                ReadState::CompressionFlag => {
+                    if this.buf.is_empty() {
+                        // Can't read compression flag right now
+                        return Ok(());
+                    } else {
+                        let compression_flag = this.buf.take(1);
+
+                        if compression_flag[0] & TRAILER_BIT == 0 {
+                            this.incomplete_data.unsplit(compression_flag);
+                            *this.state = ReadState::DataLength;
+                        } else {
+                            *this.state = ReadState::TrailerLength;
+                        }
+                    }
```

Comment:
No need to wrap this in else since the if condition achieves early return, you can reduce the nesting here.

Reply:
done